controller/lib/hal/stepper.cpp

```diff
@@ -0,0 +1,820 @@
+
+    // Get the first gate config register of the powerSTEP01.
+    // This is actually the config register on the L6470
+    mtr->GetParam(StepMtrParam::GATE_CFG1, &val);
+
+    // If this is at the default config register value for
+    // the L6470 then I don't need to do any more configuration
+    if (val == 0x2E88)
+      continue;
+
+    // Configure the two gate config registers to reasonable values
+    //
+    // GATE_CFG1 xxxxxxxxxxxxxxxx
+    //           ...........\\\\\ - time at constant current 3750ns (0x1D)
+    //           ........\\\------- Gate current 96mA (7)
+    //           .....\\\---------- Turn off boost time 1uS (7)
+    //           ....\------------- Watch dog enable (1)
+    //           \\\\-------------- reserved
+    mtr->SetParam(StepMtrParam::GATE_CFG1, 0x0FFD);
+
+    // GATE_CFG2 xxxxxxxx
+    //           ...\\\\\---------- Dead time 1000ns (7)
+    //           \\\--------------- Blanking time 1000ns (7)
+    mtr->SetParam(StepMtrParam::GATE_CFG2, 0xF7);
+  }
+}
+
+// Convert a velocity from Deg/sec units to the value to program
+// into one of the stepper controller registers
+float StepMotor::DpsToVelReg(float vel, float cnv) {
+
+  // Convert to steps / sec
+  float step_per_sec = vel * static_cast<float>(steps_per_rev_) / 360.0f;
+
+  // Multiply that by the register specific conversion factor
+  return cnv * step_per_sec;
+}
+
+// Convert a velocity from a register value to deg/sec
+float StepMotor::RegVelToDps(int32_t val, float cnv) {
+  return static_cast<float>(val) * 360.0f /
+         (cnv * static_cast<float>(steps_per_rev_));
+}
+
+// Read the current absolute motor velocity and return it
+// in deg/sec units
+// Note that this value is always positive
+StepMtrErr StepMotor::GetCurrentSpeed(float *ret) {
+  uint32_t val;
+  StepMtrErr err = GetParam(StepMtrParam::SPEED, &val);
+  *ret = RegVelToDps(val, cvt_vel_crnt_speed_reg);
+  return err;
+}
+
+// Set the motor's max speed setting in deg/sec
+//
+// NOTE - The motor must be disabled to set this
+StepMtrErr StepMotor::SetMaxSpeed(float dps) {
+  if (dps < 0)
+    return StepMtrErr::BAD_VALUE;
+
+  uint32_t speed =
+      static_cast<uint32_t>(DpsToVelReg(dps, cvt_vel_max_speed_reg));
+  if (speed > 0x3ff)
+    speed = 0x3ff;
+
+  return SetParam(StepMtrParam::MAX_SPEED, speed);
+}
+
+// Get the motor's max speed setting in deg/sec
+StepMtrErr StepMotor::GetMaxSpeed(float *ret) {
+  uint32_t val;
+  StepMtrErr err = GetParam(StepMtrParam::MAX_SPEED, &val);
+  *ret = RegVelToDps(val, cvt_vel_max_speed_reg);
+  return err;
+}
+
+// Set the motor's min speed setting in deg/sec
+//
+// NOTE - Setting a non-zero minimum speed doesn't mean
+// the motor can't stop, this is the minimum speed for
+// a move.  When you start a move, rathern then increase
+// linearly from 0, the stepper will jump to this speed
+// immediately, then ramp up from there.
+// This can help with vibration.
+//
+// NOTE - The motor must be disabled to set this
```

Comment:
Prefer not to copy-paste comments, for the same reasons we prefer not to copy-paste code.  The (very helpful) comment is in the header.

Reply:
Ya, the comments started here and I ended up moving them to the header when I decided that was a friendlier place to document how things work.
I can remove all these comments from the source if you like, but I don't think it hurts to have a short description of what the function does here.

Reply:
> I don't think it hurts to have a short description of what the function does here.

A differentiation can be that the cpp comments can speak to implementation details that wouldn't make sense in the header?  But otherwise all of the problems with copying code apply, particularly they can get out of sync and it won't be clear which one is right.